### DISCUSSION TITLE
Can I execute a bash command in yofi?

### DISCUSSION THREAD

**Discussion by pragmaticmero** [l4l/yofi#87] [Q&A]
There's a way to execute a bash command in the yofi menu?
I'm really hoping there was a way to do it.
For example: 
I hate using Super+(1-10) to cycle between window workspaces in my Tiling Window Manager (sway in my case)
So... I write a little alias command to change my workspace with a word and a number basically.
And yofi is the perfect tool to do that, just, that I don't know how to execute commands like in Dmenu. Plz help. That's a dealbreaker for me.
`a`

**Comment by l4l** [comment]
You can either:

- use `yofi binapps` mode, that one searches for binaries from `PATH` env var, so you can type smth like `swaymsg !! workspace 2`
- create a [desktop file](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s06.html) (appropriate path: `~/.local/share/applications/<any_name_of_your_file_here>.desktop`) and use the default launchmode (i.e `yofi apps`), e.g:

```
[Desktop Entry]
Name=<your entry name>
Exec=<your command goes here>
Type=Application
```

**Reply by pragmaticmero** [reply]
Well.... there's two problems, I'm going to divide this section into two sections, in one I'll describe what I'm trying to do, and the other describe the errors that I recieve.
## Sec. 1 : I'm trying to change workspaces in Sway with a bash command
The way that I've done this is:
```
# This is line 26 of my .bashrc
alias work='swaymsg workspace number'
```
Now when I type
`$ work 3`
in the command line, Sway gets me to workspace 3
The drawback using that method is that it leaves an Alacritty terminal open in every instance (I'm an Alacritty user, Rust is life <3), and I'm obligated to open a terminal instance when I want to change workspaces. So.... Its impractical in real life situations. That is why I resort to yofi, Its the best candidate for that job.

## Sec. 2 : Errors that I recieve with every method
- Method 1: yofi binapps
I was suprised that after following this method, It didn't work. Well...., I've runned "yofi binapps" in the terminal for trouble-shooting.
When i write `swaymsg !! workspace number 3` in the yofi menu, the terminal gives me:
```
$ yofi binapps
error: Found argument 'swaymsg   workspace number 3' which wasn't expected, or isn't valid in this context

USAGE:
    alacritty [FLAGS] [OPTIONS] [SUBCOMMAND]

For more information try --help
```
It appears to be an alacritty stuff.
I've run many variations of the command, but still didn't worked:
`swaymsg workspace number 3`
`"swaymsg workspace number 3"`
`alacritty !! -e swaymsg workspace number 3`
`work 3` (I had to try xd)
`bash !! -c "swaymsg workspace number 3"`
`alacritty` (Even this gives me this error)
```
error: Found argument '/usr/bin/alacritty' which wasn't expected, or isn't valid in this context

USAGE:
    alacritty [FLAGS] [OPTIONS] [SUBCOMMAND]

For more information try --help
```
(When I run `$ alacritty -e /usr/bin/alacritty` or `$ /usr/bin/alacritty` in the terminal it opens an alacritty window as expected... why is different in yofi?)
Anyways, the thing that I want to type in the Yofi menu is exactly `work <number of workspace>` , not a long-ass command LOL.

- Method 2 : create a .desktop file
This method seemed more elegant for a frontend user. I liked more this method.
**BUUUUUUT!!!!**
Crashitty crashl. Anyways, I've added an env to `$PATH` of the folder `~/.local/share/applications/` and created a .desktop file just like u said.
Yofi recognizes the app, but... when I run `work` in the menu:
```
$ yofi apps
thread 'main' panicked at 'execvp failed: ENOENT', src/exec.rs:60:41
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Yep, that's the linux life. Error tras error.
More details here:
```
$ RUST_BACKTRACE=full yofi apps
thread 'main' panicked at 'execvp failed: ENOENT', src/exec.rs:60:41
stack backtrace:
   0:     0x55c4119ce0cd - <unknown>
   1:     0x55c41188437c - <unknown>
   2:     0x55c4119cca54 - <unknown>
   3:     0x55c4119cd130 - <unknown>
   4:     0x55c4119cc52a - <unknown>
   5:     0x55c4119f3128 - <unknown>
   6:     0x55c4119f30a6 - <unknown>
   7:     0x55c4119f3062 - <unknown>
   8:     0x55c4117f93f0 - <unknown>
   9:     0x55c4117f96d2 - <unknown>
  10:     0x55c411b3ff4d - <unknown>
  11:     0x55c411b3ef22 - <unknown>
  12:     0x55c411b1ab07 - <unknown>
  13:     0x55c411b06abe - <unknown>
  14:     0x55c411b13063 - <unknown>
  15:     0x55c411af0ee9 - <unknown>
  16:     0x7ff825ffa310 - __libc_start_call_main
  17:     0x7ff825ffa3c1 - __libc_start_main@GLIBC_2.2.5
  18:     0x55c41180ba15 - <unknown>
  19:                0x0 - <unknown>
```
I don't know what's happening in either case, I'm about to throw my pc out of the window.

Thanks for the help tho, It seemed legit.

**Reply by pragmaticmero** [reply]
Extra Info:
- Arch Linux x64
- yofi v0.1.3 (I've compiled from source the version 0.1.5, still the same errors)
- Alacritty terminal 0.10.1
- Bash 5.1.16(1)

**Reply by pragmaticmero** [reply]
Quick Update: 
I'm an idiot LOL, yesterday was 4am in my country, my mind wasn't having the clarity needed for this.
```
#This is inside "work 1.desktop"
Name=work 1
Exec=swaymsg workspace number 1    #This part was inside quotes, that's why it wasn't working, fixed when I remove the quotation marks
Type=Application
```
This method works, but its kind of annoying that I have to create a .desktop file for every workspace.
There's a way to add parameters to an app in the "yofi apps" menu? The command works like this:
`$ swaymsg workspace number <number of the workspace>`
Thanks 4 the help ma dude.
I'm going to stick with this method of creating 20 .desktop files for all the workspaces that i need.

**Reply by l4l** [reply]
Sorry for a quite a delayed response.

I guess your config is missing `term` value so you should add `term = "alacritty -e"` and the first case should work. As I said before `!!` separates arguments from the command, so `swaymsg !! workspace 2` should work as-is without any desktop file (or you may use it as a shortcut).
